fl4health/clients/nnunet_client.py

```diff
@@ -0,0 +1,794 @@
+            self.steps_per_round = local_steps
+        elif local_epochs is not None:
+            self.steps_per_round = local_epochs * len(self.train_loader)
+        self.max_steps = int(config["n_server_rounds"]) * self.steps_per_round
+
+        # Create and return LR Scheduler. This is nnunet default for version 2.5.1
+        return PolyLRScheduler(
+            self.optimizers["global"], initial_lr=self.nnunet_trainer.initial_lr, max_steps=self.max_steps
+        )
+
+    def get_dataset_n_voxels(self, source_plans: Dict) -> float:
+        """
+        Determines the total number of voxels in the dataset
+
+        Args:
+            source_plans (Dict): The nnunet plans dict that is being modified
+
+        Returns:
+            float: The total number of voxels in the local client dataset
+        """
+        # Need to determine input dimensionality
+        if NnunetConfig._3D_FULLRES.value in source_plans["configurations"]:
+            cfg = source_plans["configurations"][NnunetConfig._3D_FULLRES.value]
+        else:
+            cfg = source_plans["configurations"][NnunetConfig._2D.value]
+
+        # Get total number of voxels in dataset and multiply by max_factor
+        image_shape = cfg["median_image_size_in_voxels"]
+        num_samples = self.dataset_json["numTraining"]
+        approx_n_voxels = float(np.prod(image_shape, dtype=np.float64) * num_samples)
+        return approx_n_voxels
+
+    def create_plans(self, config: Config) -> Dict[str, Any]:
+        """
+        Modifies the provided plans file to work with the local client dataset
+
+        Args:
+            config (Config): The config provided by the server. Expects the
+                'nnunet_plans' key with a pickled dictionary as the value
+
+        Returns:
+            Dict[str, Any]: The modified nnunet plans for the client
+        """
+        # Get the nnunet plans specified by the server
+        plans = pickle.loads(narrow_config_type(config, "nnunet_plans", bytes))
+
+        # Change plans name.
+        if self.plans_name is None:
+            self.plans_name = "FL_source-" + plans["plans_name"]
+
+        plans["source_plans_name"] = plans["plans_name"]
+        plans["plans_name"] = self.plans_name
+
+        # Change dataset name
+        plans["dataset_name"] = self.dataset_name
+
+        # Change data identifier and ensure batch size is within nnunet limits
+        # =====================================================================
+        if self.data_identifier is None:
+            self.data_identifier = self.plans_name
+
+        # Get maximum number of voxels for a batch based on dataset size
+        max_voxels = self.get_dataset_n_voxels(plans) * 0.05  # Max is 5% of total
+
+        # Iterate through nnunet configs in plans file
+        for c in plans["configurations"].keys():
+            # Change the data identifier
+            if "data_identifier" in plans["configurations"][c].keys():
```

Comment:
Honestly not sure, this is more just an artifact of the way nnunet was doing it. Maybe they were trying to be careful in case the user had custom trainig, optimizers etc that didn't require those keys? I don't see a reason to leave it this way as if a data_identifier is needed and one is not present that should surface an error that is pretty clear

Reply:
Gotcha. I'm alright to force `data_identifier` to be there, and maybe just log a warning if `batch_size` isn't there, as I suppose it could be hardcoded or something in a custom training process?